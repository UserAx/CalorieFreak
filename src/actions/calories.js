import database from '../firebase/firebase';

//To add new calories record
export const addCalorie = (calorie) => ({
    type: 'ADD_CALORIE',
    calorie
});

export const startAddCalorie = (calorieData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            foodname = '',
            servings = '',
            caloriesAmount = 0,
            createdAt = 0,
            description = ''
        } = calorieData;
        const calorie = {foodname, servings, caloriesAmount, createdAt, description};
        return database.ref(`users/${uid}/calories`).push(calorie).then((ref) => {
            dispatch(addCalorie({
                   id: ref.key,
                    ...calorie
                }));
            });
        }
}

//To remove calories record
export const removeCalorie = ({id} = {}) => ({
    type: 'REMOVE_CALORIE',
    id
});

export const startRemoveCalorie = ({id} = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/calories/${id}`).remove().then(() => {
            dispatch(removeCalorie({id}));
        });
    };
};

export const startEditCalorie = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/calories/${id}`).update(updates).then(() => {
            dispatch(editCalorie(id, updates));
        });
    };
};

//To edit a calories record
export const editCalorie = (id, updates) => ({
    type: 'EDIT_CALORIE',
    updates,
    id
});

export const startSetCalories = ( ) => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
      return database.ref(`users/${uid}/calories`).once('value').then((snapshot) => {
        const calories = [];
        snapshot.forEach((childSnapshot) => {
          calories.push({id: childSnapshot.key,
                        ...childSnapshot.val()});
        });
        dispatch(setCalorie(calories));
      });
    }
}

//To get a calories record from firebase to display on the page
export const setCalorie = (calories) => ({
    type: 'SET_CALORIES',
    calories
});