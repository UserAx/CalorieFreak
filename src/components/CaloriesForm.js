import React from 'react';
import {SingleDatePicker} from 'react-dates';
import moment from 'moment';

export class CaloriesForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            foodname: props.calorie ? props.calorie.foodname : '',
            servings: props.calorie ? (props.calorie.servings).toString() : '',
            caloriesAmount: props.calorie ? (props.calorie.caloriesAmount / 100).toString() : '',
            createdAt: props.calorie ? moment(props.calorie.createdAt) : moment(),
            description: props.calorie ? props.calorie.description : '',
            calanderFocus: false,
            error: undefined
        }
    }

    handleFoodNameChange = (e) => {
        const foodname = e.target.value;
        this.setState(() => ({
            foodname 
        }));
    };

    handleCaloriesChange = (e) => {
        const caloriesAmount = e.target.value;
        if(!caloriesAmount || caloriesAmount.match(/^\d{1,}(\.\d{0,2})?$/)){
        this.setState(() => ({caloriesAmount}));
        }
    };

    handleDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=> ({description}));
    };

    handleServingChange = (e) => {
        const servings = e.target.value;
        if(!servings || servings.match(/^\d{1,}?$/))
        this.setState(() => ({servings}));
    };

    handleDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({createdAt}));
        }
    };

    handleFocusChange = ({focused}) => {
        this.setState(() => ({calanderFocus: focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.foodname || !this.state.servings || !this.state.caloriesAmount){
            this.setState(() => ({error: 'Please fill in the foodname, servings and calories properly before submitting.'}));
        }else {
            this.setState(() => ({error: undefined}));
            this.props.onSubmit({
                //This function is passed from AddExpense component and edit editExpense component.
                //And here it will get the necessary values.
                foodname: this.state.foodname,
                servings: parseInt(this.state.servings),
                description: this.state.description,
                caloriesAmount: parseFloat(this.state.caloriesAmount, 10) * 100,
                //the 10 on (this.state.amount, 10), only returns the last two digits. 
                createdAt: this.state.createdAt.valueOf()
            });
        };
    }
    
    render() {
        return (
            <form className="caloriesform__container" onSubmit={this.onSubmit}>
                {this.state.error && <p className="caloriesform__error">{this.state.error}</p>}
                <input className="caloriesform__input" type="text" placeholder="What food you had today?" autoFocus 
                value={this.state.foodname} onChange={this.handleFoodNameChange}/>
                <input className="caloriesform__input" type="text" placeholder="No. of Servings" value={this.state.servings}
                onChange={this.handleServingChange}/>
                <input className="caloriesform__input"
                type="text" placeholder="I am not a psychic. Tell me how much calories you consumed."
                value={this.state.caloriesAmount} onChange={this.handleCaloriesChange}/>
                <SingleDatePicker
                date = {this.state.createdAt}
                onDateChange = {this.handleDateChange}
                focused = {this.state.calanderFocus}
                onFocusChange = {this.handleFocusChange}
                numberOfMonths = {1}
                isOutsideRange = {() => false}
                />
                <textarea className="caloriesform__textarea" placeholder="Extra notes here." 
                value={this.state.description} onChange={this.handleDescriptionChange}></textarea>
                <div className="button--addcalories">
                    <button className="button button--addcalories--extra" >Save</button>
                </div>
            </form>
        );
    }
}
export default CaloriesForm;