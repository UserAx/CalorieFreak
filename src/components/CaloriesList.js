import React from 'react';
import {connect} from 'react-redux';
import CaloriesListItem from './CaloriesListItem';
import selectCalories from '../selector/calories';

export const CaloriesList = (props) => (
    <div className="calorielist__container">
        <div className="calorielist__header">
            <div>Foods</div>
            <div>Calories</div>
        </div>
        <div className="caloriesList__items">
            {props.calories.length > 0 ?
            (
                props.calories.map((calorie) => {
                    return <CaloriesListItem key={calorie.id} {...calorie}/>
                }) 
            ) : (
                <div className="caloriesList__items__empty">
                    <span>No Records</span>
                </div>
            )}
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        calories: selectCalories(state.calories, state.filters)
    }
}

export default connect(mapStateToProps)(CaloriesList);