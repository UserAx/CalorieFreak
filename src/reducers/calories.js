const calorieDefaultState = [];

export default (state = calorieDefaultState, action) => {
    switch(action.type){
        case 'ADD_CALORIE': 
        return [
            ...state,
            action.calorie
        ];
        case 'SET_CALORIES': 
        return action.calories;
        case 'REMOVE_CALORIE':
        return state.filter(({id}) => id !== action.id);
        case 'EDIT_CALORIE':
        return state.map((calorie) => {
            if(calorie.id === action.id){
                return {
                    ...calorie,
                    ...action.updates
                }
            }else {
                return calorie;
            }
        });
        default:
        return state;
    }
}