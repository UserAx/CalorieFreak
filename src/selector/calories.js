import moment from 'moment';

export default (calories, { text, sortBy, startDate, endDate}) => {
    return calories.filter((calorie) => {
        const createdAtMoment = moment(calorie.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = calorie.foodname.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'calories'){
            return a.caloriesAmount < b.caloriesAmount ? 1 : -1;
        }else if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1;
        }
    });
}