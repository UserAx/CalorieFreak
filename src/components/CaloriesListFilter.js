import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByCalories, sortByDate, setStartDate, setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';
// import {DateRangePicker} from 'rsuite';

export class CaloriesListFilter extends React.Component {
    state = {
        calenderFocused : null
    }

    handleTextChange = (e) => {
        const text= e.target.value;
        this.props.setTextFilter(text);
    }

    handleSortByChange = (e) => {
        if(e.target.value === 'date'){
            this.props.sortByDate();
        }else if(e.target.value === 'calories'){
            this.props.sortByCalories();
        }
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calenderFocused) => {
        this.setState(() => ({calenderFocused})); 
    }

    render() {
        return (
                // <div className="burgermenu">
                //     <input type="checkbox" className="toggler" />
                //     {/* The checkbox so that we can show the menu when we click this checkbox. */}
                //     <div className="hamburger"><div></div></div>
                //     {/* THe hamburger container that will contain the actual three lines. */}
                // </div>
                
            <div className="listfilter">
                    <div className="listfilter__items">
                        <input className="listfilter__input" type="text" value={this.props.filters.text}
                        onChange={this.handleTextChange} placeholder="Filter by Text"/>
                        <select  className="listfilter__select"
                        value={this.props.filters.sortBy}
                        onChange={this.handleSortByChange}
                        >
                            <option value="date">Sort By Date</option>
                            <option value="calories">Sort By Calories</option>
                        </select>
                    </div>
                    <div>
                        <DateRangePicker
                            startDateId="MyDatePickerStartId"
                            endDateId="MyDatePickerEndId"
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates={true}/>
                    </div>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByCalories: () => dispatch(sortByCalories()),
        sortByDate: () => dispatch(sortByDate()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(CaloriesListFilter);