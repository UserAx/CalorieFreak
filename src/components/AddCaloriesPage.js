import React from 'react';
import { connect } from 'react-redux';
import CaloriesForm from './CaloriesForm';
import {startAddCalorie} from '../actions/calories';

export class AddCaloriesPage extends React.Component {    
    onSubmit = (calorie) => {
        this.props.startAddCalorie(calorie);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <div>
                    <CaloriesForm onSubmit={this.onSubmit} />
                </div>
            </div>
        )
    }
};

const mapDispatchToProps = (dispatch) => {
    return{
        startAddCalorie: (calorie) => {
            dispatch(startAddCalorie(calorie));
        }
    }
}

export default connect(undefined, mapDispatchToProps)(AddCaloriesPage);