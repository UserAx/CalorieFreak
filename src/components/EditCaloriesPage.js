import React from 'react';
import { connect } from 'react-redux';
import CaloriesForm from './CaloriesForm';
import { startEditCalorie, startRemoveCalorie } from '../actions/calories';

export class EditCaloriesPage extends React.Component{
    onSubmit = (calorie) => {
        this.props.startEditCalorie(this.props.calorie.id, calorie);
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.startRemoveCalorie({id: this.props.calorie.id});
        this.props.history.push('/');
    }
    render() {
        return (
                <div>
                    <CaloriesForm calorie = {this.props.calorie} onSubmit={this.onSubmit}/>
                    <div className="button--editcalories">
                            <button className="button button--editcalories--extra" onClick={this.onRemove}>Remove</button>
                    </div>
                </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startEditCalorie: (id, calorie) => dispatch(startEditCalorie(id, calorie)),
        startRemoveCalorie: (id) => dispatch(startRemoveCalorie(id))
    }
}

const mapStateToProps = (state, props) => {
    return {
        calorie: state.calories.find((calorie) => {return calorie.id === props.match.params.id})
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditCaloriesPage);