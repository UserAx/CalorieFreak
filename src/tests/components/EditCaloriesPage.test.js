import React from 'react';
import {shallow} from 'enzyme';
import calories from '../fixtures/calories';
import {EditCaloriesPage} from '../../components/EditCaloriesPage';

let wrapper, startEditCalorieSpy, startRemoveCalorieSpy, historySpy;
const calorie = calories[1];

beforeEach(() => {
    startEditCalorieSpy=jest.fn();
    startRemoveCalorieSpy=jest.fn();
    historySpy={push: jest.fn()};
    wrapper=shallow(<EditCaloriesPage 
        startEditCalorie={startEditCalorieSpy}
        startRemoveCalorie={startRemoveCalorieSpy} 
        history={historySpy}
        calorie={calorie}/>);
});

test('should render edit calorie form', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit action of Edit page', () => {
    wrapper.find('CaloriesForm').prop('onSubmit')(calorie);
    expect(historySpy.push).toHaveBeenCalledWith('/');
    expect(startEditCalorieSpy).toHaveBeenCalledWith(calorie.id, calorie);
});

test('should envoke onRemove on button click', () => {
    wrapper.find('button').simulate('click');
    expect(historySpy.push).toHaveBeenCalledWith('/');
    
    expect(startRemoveCalorieSpy).toHaveBeenCalledWith({id: calorie.id});
});