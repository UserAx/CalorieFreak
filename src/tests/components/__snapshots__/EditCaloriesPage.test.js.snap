// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit calorie form 1`] = `
<div>
  <div>
    <CaloriesForm
      calorie={
        Object {
          "caloriesAmount": 1504.25,
          "createdAt": -345600000,
          "description": "Snack.",
          "foodname": "Momo",
          "id": "2",
          "servings": 2,
        }
      }
      onSubmit={[Function]}
    />
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
